pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java

```diff
@@ -28,23 +29,101 @@
-  private List<FixedByteSingleValueMultiColWriter> writers = new ArrayList<>();
-  // For readers it is useful to keep a "chunked-reader" class that is aware of the start and end offset of that buffer.
-  // If we do that then we avoid the % operation on the record on every read.
-  // For the bulk read operation, we can do a loop like this one:
-  //   foreach(row in rows[]) {
-  //     bufferId = row / chunkSize;
-  //     chunked-readers.get(bufferId).getInt(row);
-  //   }
-  // TODO the logic above
-  private List<FixedByteSingleValueMultiColReader> readers = new ArrayList<>();
+  WriterWithOffset currentWriter;
+  private List<WriterWithOffset> writers = new ArrayList<>();
+  private List<ReaderWithOffset> readers = new ArrayList<>();
   private List<PinotDataBuffer> buffers = new ArrayList<>();
   private final long chunkSizeInBytes;
   private final int numRowsPerChunk;
   private final int columnSizesInBytes;
   private long numBytesInCurrentWriter = 0;
 
+  private static class WriterWithOffset implements Closeable {
+    final FixedByteSingleValueMultiColWriter writer;
+    final int startRowId;
+
+    private WriterWithOffset(FixedByteSingleValueMultiColWriter writer, int startRowId) {
+      this.writer = writer;
+      this.startRowId = startRowId;
+    }
+
+    @Override
+    public void close() {
+      writer.close();
+    }
+
+    public void setInt(int row, int value) {
+      writer.setInt(row - startRowId, 0, value);
+    }
+
+    public void setLong(int row, long value) {
+      writer.setLong(row - startRowId, 0, value);
+    }
+
+    public void setFloat(int row, float value) {
+      writer.setFloat(row - startRowId, 0, value);
+    }
+
+    public void setDouble(int row, double value) {
+      writer.setDouble(row - startRowId, 0, value);
+    }
+
+    public void setShort(int row, short value) {
+      writer.setShort(row - startRowId, 0, value);
+    }
+  }
+
+  private static class ReaderWithOffset implements Closeable {
+    final FixedByteSingleValueMultiColReader reader;
+    final int startRowId;
+
+    private ReaderWithOffset(FixedByteSingleValueMultiColReader reader, int startRowId) {
+      this.reader = reader;
+      this.startRowId = startRowId;
+    }
+
+    @Override
+    public void close() {
+      reader.close();
+    }
+
+    public int getInt(int row) {
+      return reader.getInt(row - startRowId, 0);
+    }
+
+    public long getLong(int row) {
+      return reader.getLong(row - startRowId, 0);
+    }
+
+    public float getFloat(int row) {
+      return reader.getFloat(row - startRowId, 0);
+    }
+
+    public double getDouble(int row) {
+      return reader.getDouble(row - startRowId, 0);
+    }
+
+    public short getShort(int row) {
+      return reader.getShort(row - startRowId, 0);
+    }
+
+    public void readValues(int[] rows, int rowStartPos, int rowSize, int[] values, int valuesStartPos) {
+      reader.readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
```

Comment:
Yes, that is right. I will remove these methods in favor of getting the reader and invoking it directly on the reader. Otherwise, we will need to adjust each row id into its offset.